### PR TITLE
Fix IndexOutOfRangeException in CallSite.MoveRule.

### LINKED ISSUE #36983: Race condition in concurrent method invocation with the dynamic type
### Description

Concurrent method invocation with the dynamic type from one call site intermittently causes ``System.Runtime.CompilerServices.CallSite`1.MoveRule(Int32 i)`` to throw `IndexOutOfRangeException`.  

Please check the repro project: [RaceInDynamicInvocation.zip](https://github.com/dotnet/runtime/files/4677782/RaceInDynamicInvocation.zip).
 Please invoke `build.cmd` and `run.cmd` (which continuously executes the program).

This reproduces roughly 1/20 on my environment. Please note that this issue only occurs early in the program life time.

(Example) Concurrently invoking `DynamicInvoker.Invoke` below.
```cs
public static class DynamicInvoker
{
    public static void Invoke(object o) => InvokeImpl((dynamic)o);

    private static void InvokeImpl(int x) { }
    private static void InvokeImpl(long x) { }
}
```

### Configuration

- .NET Core SDK 5.0.100-preview.4.20258.7 (x64)
- .NET Core SDK 3.1.300 (x64)
- Windows 10 Pro 1909 Build 18363.836 (x64)
- AMD Ryzen 2700X

### Regression?

Probably not. I saw this issue several times on .NET Framework 4.7.2.

### Other information

Stack Trace:

```
Unhandled exception. System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Runtime.CompilerServices.CallSite`1.MoveRule(Int32 i)
   at System.Dynamic.UpdateDelegates.UpdateAndExecuteVoid2[T0,T1](CallSite site, T0 arg0, T1 arg1)
   at RaceInDynamicInvocation.DynamicInvoker.Invoke(Object o) in C:\home\playground\RaceInDynamicInvocation\RaceInDynamicInvocation\Program.cs:line 58
   at RaceInDynamicInvocation.Program.ThreadFunc(Object obj) in C:\home\playground\RaceInDynamicInvocation\RaceInDynamicInvocation\Program.cs:line 51
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)
```

I am not really sure but my hypothesis is as follows:

- Concurrent invocation of [CallSite.AddRule](https://github.com/dotnet/runtime/blob/47ec733ba79b196e4e09d0c89bad245155002353/src/libraries/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs#L239) could shrink its `Rules` array. 
  For example:
    - Suppose `Rules` has `T[1]` (1 element).
    - Thread 1 enters `AddRule` and fetches `T[1]` from `Rules` into `rules`.
    - Thread 2 and thread 3 fully executes `AddRule` and extends `Rules`. `Rules` now has `T[3]`.
    - Thread 1 finally executes `Rules = temp`. `Rules` now has `T[2]` (shrunk!).
- `CallSite.MoveRule` could reference this shrunk array, resulting in an `IndexOutOfRangeException`.
    - `i` passed from [UpdateAndExecuteN](https://github.com/dotnet/runtime/blob/v5.0.0-preview.4.20251.6/src/libraries/System.Linq.Expressions/src/System/Dynamic/UpdateDelegates.Generated.cs#L50) could be for the extended array (`T[3]` in the above example).
    - On the other hand, when `CallSite.MoveRule` is called, `Rules` array could already be shrunk (`T[2]` in the above example).

### PR DESCRIPTION
**Pull request by asmichi**
Fix #36983

Please feel free to reject this PR if you plan to fix this issue in a different way or plan to add a concurrency test. (I failed to add a test because this fix is extremely hard to test using the public API surface...) Thank you in advance for the review!

Review perspectives:

- This fix does not hide any fundamental problems in the code. (No synchronization is the design choice.)
- The comments are correct.

### CODE REVIEW

File: src/libraries/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs [hunk 1]
```diff
@@ -266,11 +266,16 @@ internal void MoveRule(int i)
             if (i > 1)
             {
                 T[] rules = Rules!;
-                T rule = rules[i];
+                // Synchronization of AddRule is omitted for performance. Concurrent invocations of AddRule
+                // may cause Rules to revert back to an older (smaller) version, making i out of bounds.
+                if (unchecked((uint)i < (uint)rules.Length))
```
Comment (cston) [inline src/libraries/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs:271]: Please consider using `if (i < rules.Length)`.

**Comment by cston** [comment]
Thanks for fixing this @asmichi. Sorry for the delay reviewing.

**Comment by cston** [comment]
> I failed to add a test because this fix is extremely hard to test using the public API surface

I've created a test that calls `CallSite<T>.AddRule()` and `MoveRule()` directly (see https://github.com/dotnet/runtime/pull/40858). It hits the race condition reliably without your fix.

File: src/libraries/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs (hunk 1 above)
Comment (asmichi) [reply src/libraries/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs]: Simplified to `if (i < rules.Length)`. Thanks!

**Review by cston — APPROVED** [review]

**Comment by asmichi** [comment]
https://dev.azure.com/dnceng/public/_build/results?buildId=772308&view=results (dotnet-linker-tests)
> ##[error]The job running on agent Azure Pipelines 11 ran longer than the maximum time of 120 minutes. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134

Looks like this job intermittently times out?
https://dev.azure.com/dnceng/public/_build?definitionId=861&statusFilter=failed (Many >2h runs)

**Review by 333fred — APPROVED** [review]

**Comment by cston** [comment]
Thanks @asmichi.

**Pull request merged**